doc/developer/design/20231211_datadog_sink.md

```diff
@@ -0,0 +1,162 @@
+
+While we could allow the user to set the timestamp, it would make the experience harder. Collections like `mz_cluster_replica_utilization` do not include a timestamp, forcing users to create one themselves. *I do not discard the chance to make it optional.*
+
+### Materialize source structure
+
+A Datadog sink needs a _source_: a table or view — and must have a structure that matches [Datadog's API submit metrics endpoint](https://docs.datadoghq.com/api/latest/metrics/#submit-metrics) fields.
+
+A table in Materialize acting as a source for the sink must have the following structure:
+
+```sql
+CREATE TABLE metrics (
+    -- Required:
+    metric TEXT,
+    resources JSONB,
+    value FLOAT,
+    -- Optional:
+    type INT,
+    interval BIGINT,
+    tags TEXT[],
+    unit TEXT,
+    metadata: JSONB,
+);
+```
+
+And a view needs the same structure:
+```sql
+CREATE MATERIALIZED VIEW metrics AS
+    SELECT
+    -- Required:
+    DISTINCT(
+        jsonb_build_object(
+            'cluster_id', R.cluster_id,
+            'cluster_name', C.name,
+        )
+    ) as resources,
+    'cluster_replica_cpu_usage' as metric,
+    cpu_percent::float as value,
+    -- Optional:
+    0 as type,
+        120 as interval,
+    ARRAY['production', 'Materialize'] as tags,
+    'percent' as unit
+    FROM mz_internal.mz_cluster_replica_utilization U
+    JOIN mz_catalog.mz_cluster_replicas R ON (U.replica_id = R.id)
+    JOIN mz_catalog.mz_clusters C ON (R.cluster_id = C.id);
+```
+
+**NOTE**: Metrics and their values do not need to be unique for each timestamp, but the latest value will suppress the other in Datadog. It would be great if the sink could detect this during creation/execution and emit a warning if there are different values at the same point in time for a set of metrics and resources.
+
+Another distinction is that *resources* equal *labels* in the OpenMetrics definition. I kept *resources* to match Datadog’s API, but I like *labels* as a possible replacement.
+
+### OAuth Process
+
+Datadog recommends an OAuth process for their integrations and creating an API key on behalf of the user if the integration needs to write data into Datadog. The token does not have the scope to write data into Datadog, so creating the API key is necessary.
+
+The OAuth process requires additional logic in the console. The whole process is defined in the following sequence diagram:
+
+<details open>
+<summary>Sequence diagram</summary>
+<img width="2129" alt="Sequence Diagram" src="https://github.com/joacoc/materialize/assets/11491779/28dc14cb-a488-4e8b-8c89-ab69accfd858">
+</details>
+
+At the end of the flow, the console creates a secret containing the Datadog API key, connection, and sink. Although we lack an integrations page, a screen similar to what the mz displays at the end of the flow is more than enough for an MVP.
+
```

Comment:
As I mentioned on Slack, this seems like a ton of rigamarole that Datadog has suggested, just for us to end up with an API key! As long as Datadog isn't banning users from creating API keys any time soon, I think we should just tell users to create an API key by hand and install it as a secret in Materialize. That will save us a lot of time, and we can revisit the OAuth road at a future date when we have more data about how many users are using the Datadog integration.